roles/test-utils/mining-device/src/main.rs

```diff
@@ -499,3 +610,43 @@ impl Miner {
         }
     }
 }
+
+// returns hashrate based on how fast the device hashes over the given duration
+fn measure_hashrate(duration_secs: u64) -> f64 {
+    let mut rng = thread_rng();
+    let prev_hash: [u8; 32] = generate_random_32_byte_array().to_vec().try_into().unwrap();
+    let prev_hash = Hash::from_inner(prev_hash);
+    let merkle_root: [u8; 32] = generate_random_32_byte_array().to_vec().try_into().unwrap();
+    let merkle_root = Hash::from_inner(merkle_root);
+    let header = BlockHeader {
+        version: rng.gen(),
+        prev_blockhash: BlockHash::from_hash(prev_hash),
+        merkle_root,
+        time: std::time::SystemTime::now()
+            .duration_since(std::time::SystemTime::UNIX_EPOCH - std::time::Duration::from_secs(60))
+            .unwrap()
+            .as_secs() as u32,
+        bits: rng.gen(),
+        nonce: 0,
+    };
+    let start_time = Instant::now();
+    let mut hashes: u64 = 0;
+    let duration = Duration::from_secs(duration_secs);
+    let mut miner = Miner::new(0);
+    miner.new_target(vec![0_u8; 32]);
+    miner.header = Some(header);
+
+    while start_time.elapsed() < duration {
+        let _ = miner.next_share();
+        hashes += 1;
+    }
```

Comment:
why are we setting the target as 0?

doesn't this make it basically impossible to find a share?
```
difficulty = genesis_target / target = genesis_target / 0 = ∞
```

we are bypassing this restriction because we are ignoring the error returned by `miner.next_share()`, but that makes the code more confusing and counter-intuitive than it needs to be

but in order to improve readability and make a more clear usage of the API provided by `miner.next_share()`, I would refactor this snippet into something like:

```suggestion
    // set the max possible target (genesis hash), so that every share is valid
    let max_target = "00000000ffff0000000000000000000000000000000000000000000000000000";
    let target = hex::decode(max_target).expect("Decoding failed");
    miner.new_target(target);
    miner.header = Some(header);

    while start_time.elapsed() < duration {
        miner.next_share().expect("every share is valid under max_target");
        hashes += 1;
    }
```

but when I test these changes, I get a panic:
```
2024-05-03T20:48:04.540935Z  INFO mining_device: Set target to 0000000000000000000000000000000000000000000000000000ffff00000000
thread 'main' panicked at test-utils/mining-device/src/main.rs:644:28:
every share is valid under max_target: ()
```

which means the implementation of `fn next_share` is probably broken, probably on [this line](https://github.com/Fi3/stratum-1/blob/e4ad646c344d7ea124fbf651c575e4b956940bc3/roles/test-utils/mining-device/src/main.rs#L602):
```
if hash < *self.target.as_ref().ok_or(())?
```

Reply:
this log line is also concerning:
```
2024-05-03T20:48:04.540935Z  INFO mining_device: Set target to 0000000000000000000000000000000000000000000000000000ffff00000000
```

why is the target hash being displayed backwards? maybe an indication of the root cause of the bug?